core/src/main/java/tech/tablesaw/io/xlsx/XlsxReader.java

```diff
@@ -0,0 +1,312 @@
+        final List<String> headerNames = new ArrayList<>();
+        for (final Cell cell : row) {
+            if (cell.getCellType() == CellType.STRING) {
+                headerNames.add(cell.getRichStringCellValue().getString());
+            } else {
+                break;
+            }
+        }
+        if (headerNames.size() == tableArea[3] - tableArea[2] + 1) {
+            tableArea[0]++;
+        } else {
+            headerNames.clear();
+            for (int col = tableArea[2]; col <= tableArea[3]; col++) {
+                headerNames.add("col" + col);
+            }
+        }
+        final Table table = Table.create(options.tableName());
+        final List<Column<?>> columns = new ArrayList<>();
+        while (columns.size() < headerNames.size()) {
+            columns.add(null);
+        }
+        for (int rowNum = tableArea[0]; rowNum <= tableArea[1]; rowNum++) {
+            row = sheet.getRow(rowNum);
+            for (int colNum = 0; colNum < headerNames.size(); colNum++) {
+                final Cell cell = row.getCell(colNum + tableArea[2], MissingCellPolicy.RETURN_BLANK_AS_NULL);
+                Column<?> column = columns.get(colNum);
+                if (cell != null) {
+                    if (column == null) {
+                        column = createColumn(headerNames.get(colNum), cell);
+                        columns.set(colNum, column);
+                        while (column.size() < rowNum - tableArea[0]) {
+                            column.appendMissing();
+                        }
+                    }
+                    final Column<?> altColumn = appendValue(column, cell);
+                    if (altColumn != null && altColumn != column) {
+                        column = altColumn;
+                        columns.set(colNum, column);
+                    }
+                }
+                if (column != null) {
+                	while (column.size() <= rowNum - tableArea[0]) {
+                		column.appendMissing();
+                	}
+                }
+            }
+        }
+        while (columns.remove(null));
+        table.addColumns(columns.toArray(new Column<?>[columns.size()]));
+        return table;
+    }
+
+    private Column<?> appendValue(final Column<?> column, final Cell cell) {
+        switch (cell.getCellType()) {
+        case STRING:
+            column.appendCell(cell.getRichStringCellValue().getString());
+            return null;
+        case NUMERIC:
+            if (DateUtil.isCellDateFormatted(cell)) {
+                final Date date = cell.getDateCellValue();
+                final LocalDateTime localDate = LocalDateTime.of(date.getYear() + 1900, date.getMonth() + 1, date.getDate(), date.getHours(), date.getMinutes(), date.getSeconds());
+                column.appendCell(localDate.toString());
+                return null;
+            } else {
+                final double num = cell.getNumericCellValue();
```

Comment:
In the other readers we use `ColumnTypeDetector` to figure out the column type upfront and then don't need the logic at append time. I wonder if you could do the same here?

Reply:
I saw **ColumnTypeDetector**, but it doesn't quite fit my use case, since I have objects, not strings. I can reimplement the same strategy, i.e. scanning a number of elements in the columns and selecting the appropriate **ColumnType**. Is it OK that I extend ColumnType with a **supports(Object)** method to help selecting the appropriate type (similar to the parser's **canParse** method)?

Reply:
Might as well just leave it as is for now

I didn't review the PR too much cause it was hard to read with the mixture of tabs and spaces. I'll give it a closer look after that's updated. But it seems pretty good overall